Fixups to timer_api test

andyross opened the pull request:
This test sets a timer, busy waits for half the duration, and then
checks the remaining time is correct.  And it correctly does all its
math in tick precision and aligns to a timer interrupt to eliminate
aliasing due to the tick stride.

But it's waiting using k_busy_wait(), not a timer: "half the duration"
in MICROSECONDS (for k_busy_wait()) is not necessarily representable
as an integer number of TICKS on all platforms.  Because k_busy_wait()
always rounds up, we need one extra tick of buffer on those platforms.

Fixes #24001

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/timer/timer_api: Correct tick math for odd freq timers

This test sets a timer, busy waits for half the duration, and then
checks the remaining time is correct.  And it correctly does all its
math in tick precision and aligns to a timer interrupt to eliminate
aliasing due to the tick stride.

But it's waiting using k_busy_wait(), not a timer: "half the duration"
in MICROSECONDS (for k_busy_wait()) is not necessarily representable
as an integer number of TICKS on all platforms.  Because k_busy_wait()
always rounds up, we need one extra tick of buffer on those platforms.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
@dcpleung reported this on the microchip board.  I haven't tested it there, and (oddly) it doesn't seem to happen on my nRF board which should have roughly the same properties.  He should validate before merge.

andyross commented:
(FWIW: I just remembered that nRF are 1:1 with ticks:cycles, so they aren't subject to this problem.  We should check, if the microchip config has a non-divisible ratio we should fix that because it can cause problems like this in app code.)

dcpleung reviewed (approved):
This fixed the `test_timer_remaining()` test on `mec15xxevb_assy6853`. (relates to #24001)

pabigot commented:
While Nordic has ticks and cycles the same, the busywait uses a different clock.  Further, on it an integral number of ticks is rarely an exact number of milliseconds.  So this may also be a factor in #23106.

nashif commented:
just ran this on a few boards:
```
junit2html sanity-out/*.xml  --summary-matrix

Matrix Test Report
===================
                                                  disco_l475_iot1.xml
                                                  | frdm_k64f.xml
                                                  | | hifive1_revb.xml
                                                  | | | mimxrt1050_evk.xml
                                                  | | | | nrf52_pca10040.xml
                                                  | | | | | nucleo_f207zg.xml
                                                  | | | | | | reel_board.xml
                                                  | | | | | | | sam_e70_xplained.xml
                                                  | | | | | | | |
disco_l475_iot1:kernel.timer
- kernel.timer.time_conversions                   /                / Passed
- kernel.timer.timeout_abs                        /                / Passed
- kernel.timer.timer_duration_period              /                / Passed
- kernel.timer.timer_expirefn_null                /                / Passed
- kernel.timer.timer_k_define                     f                F Total Failure
- kernel.timer.timer_period_0                     /                / Passed
- kernel.timer.timer_periodicity                  /                / Passed
- kernel.timer.timer_remaining                    /                / Passed
- kernel.timer.timer_status_get                   /                / Passed
- kernel.timer.timer_status_get_anytime           /                / Passed
- kernel.timer.timer_status_sync                  /                / Passed
- kernel.timer.timer_user_data                    /                / Passed
disco_l475_iot1:kernel.timer.tickless
- kernel.timer.tickless.time_conversions          /                / Passed
- kernel.timer.tickless.timeout_abs               /                / Passed
- kernel.timer.tickless.timer_duration_period     /                / Passed
- kernel.timer.tickless.timer_expirefn_null       /                / Passed
- kernel.timer.tickless.timer_k_define            /                / Passed
- kernel.timer.tickless.timer_period_0            /                / Passed
- kernel.timer.tickless.timer_periodicity         /                / Passed
- kernel.timer.tickless.timer_remaining           /                / Passed
- kernel.timer.tickless.timer_status_get          /                / Passed
- kernel.timer.tickless.timer_status_get_anytime  /                / Passed
- kernel.timer.tickless.timer_status_sync         /                / Passed
- kernel.timer.tickless.timer_user_data           /                / Passed
frdm_k64f:kernel.timer
- kernel.timer.time_conversions                     /              / Passed
- kernel.timer.timeout_abs                          /              / Passed
- kernel.timer.timer_duration_period                /              / Passed
- kernel.timer.timer_expirefn_null                  /              / Passed
- kernel.timer.timer_k_define                       /              / Passed
- kernel.timer.timer_period_0                       /              / Passed
- kernel.timer.timer_periodicity                    /              / Passed
- kernel.timer.timer_remaining                      /              / Passed
- kernel.timer.timer_status_get                     /              / Passed
- kernel.timer.timer_status_get_anytime             /              / Passed
- kernel.timer.timer_status_sync                    /              / Passed
- kernel.timer.timer_user_data                      /              / Passed
frdm_k64f:kernel.timer.tickless
- kernel.timer.tickless.time_conversions            /              / Passed
- kernel.timer.tickless.timeout_abs                 /              / Passed
- kernel.timer.tickless.timer_duration_period       /              / Passed
- kernel.timer.tickless.timer_expirefn_null         /              / Passed
- kernel.timer.tickless.timer_k_define              /              / Passed
- kernel.timer.tickless.timer_period_0              /              / Passed
- kernel.timer.tickless.timer_periodicity           /              / Passed
- kernel.timer.tickless.timer_remaining             /              / Passed
- kernel.timer.tickless.timer_status_get            /              / Passed
- kernel.timer.tickless.timer_status_get_anytime    /              / Passed
- kernel.timer.tickless.timer_status_sync           /              / Passed
- kernel.timer.tickless.timer_user_data             /              / Passed
hifive1_revb:kernel.timer
- kernel.timer.time_conversions                       f            F Total Failure
- kernel.timer.timeout_abs                            f            F Total Failure
- kernel.timer.timer_duration_period                  f            F Total Failure
- kernel.timer.timer_expirefn_null                    f            F Total Failure
- kernel.timer.timer_k_define                         f            F Total Failure
- kernel.timer.timer_period_0                         f            F Total Failure
- kernel.timer.timer_periodicity                      f            F Total Failure
- kernel.timer.timer_remaining                        f            F Total Failure
- kernel.timer.timer_status_get                       f            F Total Failure
- kernel.timer.timer_status_get_anytime               f            F Total Failure
- kernel.timer.timer_status_sync                      f            F Total Failure
- kernel.timer.timer_user_data                        f            F Total Failure
mimxrt1050_evk:kernel.timer
- kernel.timer.time_conversions                         /          / Passed
- kernel.timer.timeout_abs                              /          / Passed
- kernel.timer.timer_duration_period                    /          / Passed
- kernel.timer.timer_expirefn_null                      /          / Passed
- kernel.timer.timer_k_define                           /          / Passed
- kernel.timer.timer_period_0                           /          / Passed
- kernel.timer.timer_periodicity                        /          / Passed
- kernel.timer.timer_remaining                          /          / Passed
- kernel.timer.timer_status_get                         /          / Passed
- kernel.timer.timer_status_get_anytime                 /          / Passed
- kernel.timer.timer_status_sync                        /          / Passed
- kernel.timer.timer_user_data                          /          / Passed
mimxrt1050_evk:kernel.timer.tickless
- kernel.timer.tickless.time_conversions                /          / Passed
- kernel.timer.tickless.timeout_abs                     /          / Passed
- kernel.timer.tickless.timer_duration_period           /          / Passed
- kernel.timer.tickless.timer_expirefn_null             /          / Passed
- kernel.timer.tickless.timer_k_define                  /          / Passed
- kernel.timer.tickless.timer_period_0                  /          / Passed
- kernel.timer.tickless.timer_periodicity               /          / Passed
- kernel.timer.tickless.timer_remaining                 /          / Passed
- kernel.timer.tickless.timer_status_get                /          / Passed
- kernel.timer.tickless.timer_status_get_anytime        /          / Passed
- kernel.timer.tickless.timer_status_sync               /          / Passed
- kernel.timer.tickless.timer_user_data                 /          / Passed
nrf52_pca10040:kernel.timer
- kernel.timer.time_conversions                           /        / Passed
- kernel.timer.timeout_abs                                f        F Total Failure
- kernel.timer.timer_duration_period                      /        / Passed
- kernel.timer.timer_expirefn_null                        /        / Passed
- kernel.timer.timer_k_define                             /        / Passed
- kernel.timer.timer_period_0                             /        / Passed
- kernel.timer.timer_periodicity                          /        / Passed
- kernel.timer.timer_remaining                            f        F Total Failure
- kernel.timer.timer_status_get                           /        / Passed
- kernel.timer.timer_status_get_anytime                   /        / Passed
- kernel.timer.timer_status_sync                          /        / Passed
- kernel.timer.timer_user_data                            /        / Passed
nrf52_pca10040:kernel.timer.tickless
- kernel.timer.tickless.time_conversions                  /        / Passed
- kernel.timer.tickless.timeout_abs                       f        F Total Failure
- kernel.timer.tickless.timer_duration_period             /        / Passed
- kernel.timer.tickless.timer_expirefn_null               /        / Passed
- kernel.timer.tickless.timer_k_define                    /        / Passed
- kernel.timer.tickless.timer_period_0                    /        / Passed
- kernel.timer.tickless.timer_periodicity                 /        / Passed
- kernel.timer.tickless.timer_remaining                   f        F Total Failure
- kernel.timer.tickless.timer_status_get                  /        / Passed
- kernel.timer.tickless.timer_status_get_anytime          /        / Passed
- kernel.timer.tickless.timer_status_sync                 /        / Passed
- kernel.timer.tickless.timer_user_data                   /        / Passed
nucleo_f207zg:kernel.timer
- kernel.timer.time_conversions                             /      / Passed
- kernel.timer.timeout_abs                                  /      / Passed
- kernel.timer.timer_duration_period                        /      / Passed
- kernel.timer.timer_expirefn_null                          /      / Passed
- kernel.timer.timer_k_define                               /      / Passed
- kernel.timer.timer_period_0                               /      / Passed
- kernel.timer.timer_periodicity                            /      / Passed
- kernel.timer.timer_remaining                              /      / Passed
- kernel.timer.timer_status_get                             /      / Passed
- kernel.timer.timer_status_get_anytime                     /      / Passed
- kernel.timer.timer_status_sync                            /      / Passed
- kernel.timer.timer_user_data                              /      / Passed
nucleo_f207zg:kernel.timer.tickless
- kernel.timer.tickless.time_conversions                    /      / Passed
- kernel.timer.tickless.timeout_abs                         /      / Passed
- kernel.timer.tickless.timer_duration_period               /      / Passed
- kernel.timer.tickless.timer_expirefn_null                 /      / Passed
- kernel.timer.tickless.timer_k_define                      /      / Passed
- kernel.timer.tickless.timer_period_0                      /      / Passed
- kernel.timer.tickless.timer_periodicity                   /      / Passed
- kernel.timer.tickless.timer_remaining                     /      / Passed
- kernel.timer.tickless.timer_status_get                    /      / Passed
- kernel.timer.tickless.timer_status_get_anytime            /      / Passed
- kernel.timer.tickless.timer_status_sync                   /      / Passed
- kernel.timer.tickless.timer_user_data                     /      / Passed
reel_board:kernel.timer
- kernel.timer.time_conversions                               /    / Passed
- kernel.timer.timeout_abs                                    f    F Total Failure
- kernel.timer.timer_duration_period                          /    / Passed
- kernel.timer.timer_expirefn_null                            /    / Passed
- kernel.timer.timer_k_define                                 /    / Passed
- kernel.timer.timer_period_0                                 /    / Passed
- kernel.timer.timer_periodicity                              /    / Passed
- kernel.timer.timer_remaining                                /    / Passed
- kernel.timer.timer_status_get                               /    / Passed
- kernel.timer.timer_status_get_anytime                       /    / Passed
- kernel.timer.timer_status_sync                              /    / Passed
- kernel.timer.timer_user_data                                /    / Passed
reel_board:kernel.timer.tickless
- kernel.timer.tickless.time_conversions                      /    / Passed
- kernel.timer.tickless.timeout_abs                           f    F Total Failure
- kernel.timer.tickless.timer_duration_period                 /    / Passed
- kernel.timer.tickless.timer_expirefn_null                   /    / Passed
- kernel.timer.tickless.timer_k_define                        /    / Passed
- kernel.timer.tickless.timer_period_0                        /    / Passed
- kernel.timer.tickless.timer_periodicity                     /    / Passed
- kernel.timer.tickless.timer_remaining                       /    / Passed
- kernel.timer.tickless.timer_status_get                      /    / Passed
- kernel.timer.tickless.timer_status_get_anytime              /    / Passed
- kernel.timer.tickless.timer_status_sync                     /    / Passed
- kernel.timer.tickless.timer_user_data                       /    / Passed
sam_e70_xplained:kernel.timer
- kernel.timer.time_conversions                                 /  / Passed
- kernel.timer.timeout_abs                                      /  / Passed
- kernel.timer.timer_duration_period                            /  / Passed
- kernel.timer.timer_expirefn_null                              /  / Passed
- kernel.timer.timer_k_define                                   /  / Passed
- kernel.timer.timer_period_0                                   /  / Passed
- kernel.timer.timer_periodicity                                /  / Passed
- kernel.timer.timer_remaining                                  /  / Passed
- kernel.timer.timer_status_get                                 /  / Passed
- kernel.timer.timer_status_get_anytime                         /  / Passed
- kernel.timer.timer_status_sync                                /  / Passed
- kernel.timer.timer_user_data                                  /  / Passed
sam_e70_xplained:kernel.timer.tickless
- kernel.timer.tickless.time_conversions                        /  / Passed
- kernel.timer.tickless.timeout_abs                             /  / Passed
- kernel.timer.tickless.timer_duration_period                   /  / Passed
- kernel.timer.tickless.timer_expirefn_null                     /  / Passed
- kernel.timer.tickless.timer_k_define                          /  / Passed
- kernel.timer.tickless.timer_period_0                          /  / Passed
- kernel.timer.tickless.timer_periodicity                       /  / Passed
- kernel.timer.tickless.timer_remaining                         /  / Passed
- kernel.timer.tickless.timer_status_get                        /  / Passed
- kernel.timer.tickless.timer_status_get_anytime                /  / Passed
- kernel.timer.tickless.timer_status_sync                       /  / Passed
- kernel.timer.tickless.timer_user_data                         /  / Passed

-------------------------------------------------------------------------------
Test Results:
  Failed       :     19
  Passed       :    161
```

Commit:
tests/kernel/timer/timer_api: Fix static double-conversion goof

The test of the absolute timeout feature was a simple whitebox test
that inspected the generated ticks field of different constructors for
identity.  But it wasn't simple enough, because it was doing a
ticks->ms->ticks conversion (at compile time, sigh) on the input data,
which is obviously lossy on platforms where ticks are shorter than
milliseconds by non-integral factors.

Fix to do the conversion in just one direction.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Add a patch to fix the abs_timeout test's conversion goof.

andyross commented:
Rebase to clear out that weird doc failure.

FRASTM commented:
@andyross , I see your change to add one tick to the`  zassert_true(((s64_t)exp_ticks - (s64_t)now) <= (dur_ticks / 2) + 1, NULL);`
My question was about the previous check :  on our side, that 
    zassert_true(rem_ticks <= (dur_ticks / 2) **+ 1** , NULL);  
is fixing same kind of error. 
Why this one-tick delay you could not apply there ?

FRASTM commented:
Fixes #24072

andrewboie commented:
@nashif does this still cause regressions, or is it now an improvement on the master branch with your testing rig?
